src/main/java/ch/njol/skript/expressions/ExprCraftingSlots.java

```diff
@@ -0,0 +1,171 @@
+	"on preparing crafting:",
+	"\tset crafting result item to wood"})
+@Since("INSERT VERSION")
+public class ExprCraftingSlots extends SimpleExpression<ItemStack> {
+	
+	static {
+		Skript.registerExpression(ExprCraftingSlots.class, ItemStack.class, ExpressionType.COMBINED,
+				"[the] crafting [inventory] result (slot|item) [of %inventories%]",
+				"%inventories%'[s] crafting [inventory] result (slot|item)",
+				"[the] crafting [inventory] (matrix|grid) [(slots|items|shape)] [of %inventories%]",
+				"%inventories%'[s] crafting [inventory] (matrix|grid) [(slots|items|shape)]");
+	}
+
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private Expression<Inventory> invis;
+	private boolean isMatrix;
+	final ItemStack AIR = new ItemStack(Material.AIR);
+	final ItemStack[] AIR_ITEMSTACK_LIST = new ItemStack[]{AIR, AIR, AIR, AIR, AIR, AIR, AIR, AIR, AIR};
+
+	@Override
+	@SuppressWarnings({"null", "unchecked"})
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		invis = (Expression<Inventory>) exprs[0];
+		isMatrix = matchedPattern > 1;
+		return true;
+	}
+
+	@Override
+	@Nullable
+	protected ItemStack[] get(Event e) {
+		Inventory[] invis = this.invis.getArray(e);
+		if (invis == null)
+			return null;
+
+		List<ItemStack> items = new ArrayList<>(0);
+		for (Inventory invi : invis) {
+			if (!(invi instanceof CraftingInventory))
+				continue;
+
+			CraftingInventory craft = (CraftingInventory) invi;
+			if (isMatrix) {
+				items.addAll(Arrays.asList(craft.getMatrix()));
+			} else {
+				items.add(craft.getResult());
+			}
+		}
+		return items.toArray(new ItemStack[0]);
+	}
+
+	@Override
+	public @Nullable Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+			case SET:
+			case DELETE:
+				return CollectionUtils.array(ItemStack[].class);
+			default:
+				return null;
+		}
+	}
+
+	@Override
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		Inventory[] invis = this.invis.getArray(e);
+		ItemStack[] items = (ItemStack[]) delta;
+		if (invis == null)
+			return;
+
+		if (e instanceof PrepareItemCraftEvent && isMatrix) { // Until spigot/paper fixes this
+			Skript.warning("Editing the crafting inventory matrix in player " +
+				"preparing craft event will result in an infinite loop due " +
+				"to how spigot/paper handle this event, therefore it's disabled for now.");
+			return;
+		}
+
+		if (mode == ChangeMode.DELETE) {
+			for (Inventory invi : invis) {
+				if (!(invi instanceof CraftingInventory))
+					continue;
+				if (isMatrix)
+					((CraftingInventory) invi).setMatrix(AIR_ITEMSTACK_LIST);
+				else
+					((CraftingInventory) invi).setResult(new ItemStack(Material.AIR));
+			}
+		} else { // SET
+			assert invis != null;
+			if (items == null)
+				return;
+			for (Inventory invi : invis) {
+				if (!(invi instanceof CraftingInventory))
+					continue;
+				if (isMatrix) {
+					List<ItemStack> itemStacks = new ArrayList<>(9);
```

Comment:
We could just be using an Array here over an ArrayList as ArrayLists are designed to be expanded based on an unknown amount of values. Here we know it's 9.

This will also improve performance using Arrays over ArrayLists.